### ISSUE TITLE
Add ability to see chat logs across shutdowns

### ISSUE THREAD

**Issue by tandy-1000** [slskd/slskd#211]
At the moment, when slskd is restarted, chat logs are lost. This might be best implemented with a config option to remove chats after `x` number of days .

**Comment by neofright** [comment]
I'd also love to see this feature.

Presumably the chats are stored in RAM. Is there any command that could be run in the container to save chats to a .txt in the interim?

**Comment by neofright** [comment]
```
#!/usr/bin/env bash
binary_name="slskd"
pid="$(pgrep "$binary_name")"

new_messages=()
if [[ -n "$pid" ]];
then
    while IFS="-" read -r a b 
    do
	of_file="${pid}_mem_${a}.bin"
        dd if="/proc/${pid}/mem" bs="$(getconf PAGESIZE)" iflag=skip_bytes,count_bytes \
        skip=$(( 0x$a )) count=$(( 0x$b - 0x$a )) of="$of_file" 2>/dev/null

	while read -r mmap_message; do new_messages+=("$mmap_message"); done < <(grep -a '"acknowledged":false' "$of_file"|sort -u)

	rm "$of_file" 
    done < <(grep -E "^([0-9a-f]*)-([0-9a-f]*) rw-p ([0-9]*) ([0-9]*:[0-9]*) 0 " "/proc/${pid}/maps"|grep -vF '.so'|awk '{print $1}')
fi
for message in "${new_messages[@]}"; do echo "$message" | jq; done
```
Information taken from the following URLs:
https://serverfault.com/questions/173999/dump-a-linux-processs-memory-to-file
https://davidebove.com/blog/2021/03/27/how-to-dump-process-memory-in-linux/
https://www.baeldung.com/linux/proc-id-maps

- grep for memory addresses:
 "rw-p" signifies data and not code.
 inode ids of '0' mean the memory is not a mapped file.
 crude inverse match to remove shared object libraries.
- Read the memory address's start as 'a' and the address's end as 'b'
- Read all unique and unread json messages contained in this memory segment into the array.
- Remove the memory segment from disk after scraping it for messages.
- Print our unread messages and pipe them to jq for pretty printing.

The script has to be run as root. Below is a redacted sample output of the script:
```
root@HOSTNAME:# ./slskd_mem_dump.sh
{
  "user_name": [
    {
      "acknowledged": false,
      "id": 12345678,
      "message": "Hey, nice script buddy!",
      "replayed": false,
      "timestamp": "2022-12-20T23:12:33Z",
      "username": "user_name"
    }
  ]
}
```

I'm going to throw this as a cron job on my server, so it will send me an email when I have a new message to read. Once I check my messages it will mark as `"acknowledged": true` and I will not be sent the same message again via email.

Is this a disgusting kludge? Yes.

Currently it dumps the entire SQL database to disk as it resides in memory. I don't know of a more efficient way to extract only this data...

**Comment by jpdillingham** [comment]
Impressive!  I'm going to be working on this feature this week so this shouldn't be necessary.

I think an email integration similar to the existing one for Pushbullet would be fairly easy to pull off, also.

**Comment by jpdillingham** [comment]
Included in release [0.16.35](https://github.com/slskd/slskd/releases/tag/0.16.35) 🎉
